### DISCUSSION TITLE
Is there a warnings-by-level list anywhere?

### DISCUSSION THREAD

**Discussion by SRNissen** [dotnet/roslyn#71770] [Q&A]
I have a project with a CS8618 warning shown by my IDE (JetBrains Rider).

The warning is correct (that is, the field is not initialized before leaving the ctor)

The .csproj has:

    <PropertyGroup Condition=" '$(Configuration)' == 'Debug' ">
        <AnalysisLevel>latest</AnalysisLevel>
        <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
    </PropertyGroup>

    <PropertyGroup Condition=" '$(Configuration)' == 'Release' ">
        <AnalysisLevel>latest</AnalysisLevel>
        <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
    </PropertyGroup>

I have tried with any combination of Warning and Analysis level but I *cannot* make the project stop building.

I've been looking through a lot of documentation over the last few hours and I simply cannot find a list of warnings-by-warning-level that'll tell me how my Rider setup is different from my project and, in particular, what to do to make this project stop building on this and similar errors.

**Comment by huoyaoyuan** [comment]
https://learn.microsoft.com/dotnet/csharp/language-reference/compiler-messages/warning-waves

CS8618 is a [nullable warning](https://learn.microsoft.com/dotnet/csharp/language-reference/compiler-messages/nullable-warnings#nonnullable-reference-not-initialized). Do you have `<Nullable>enable</Nullable>` in your project? Is there warning reported when doing command line build?

**Reply by SRNissen** [reply]
I am aware CS8618 is a nullable warning. The linked article does not say which warning wave it belongs to.

**Reply by huoyaoyuan** [reply]
It seems that it does not belongs to a wave. It belongs to C# 8.
Can you check the warning in command line build?

**Reply by SRNissen** [reply]
Sure - I have a public repo that demos the issue


```sh
$ gh repo clone SRNissen/demos

Cloning into 'demos'...
remote: Enumerating objects: 20, done.
remote: Counting objects: 100% (20/20), done.
remote: Compressing objects: 100% (15/15), done.
remote: Total 20 (delta 5), reused 19 (delta 4), pack-reused 0
Receiving objects: 100% (20/20), 6.17 KiB | 6.17 MiB/s, done.
Resolving deltas: 100% (5/5), done.

$ cd demos/warninglist

$ dotnet build

MSBuild version 17.8.3+195e7f5a3 for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
  warninglist -> /home/srnissen/repos/srnissen/demos/warninglist/bin/Debug/net8.0/warninglist.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:01.06
```

- To be clear: With this question,
  - I **am not** looking for:
    - Changes to this project.
  - I **am** looking for:
    - C# documentation of all warnings, and the flags necessary to enable/disable them.

**Reply by huoyaoyuan** [reply]
Most warnings don't belong to a warning wave. They just belong to the language version they were introduced in. Most nullable warnings belong to C# 8. Only warning for previously legal code would be included in a warning wave, to avoid sudden change for compiler version upgrade.

Look at the source code about warning level:

https://github.com/dotnet/roslyn/blob/983d20f09579f435adf3b07af9017ca9c64e61c0/src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs#L206-L241

So [this page](https://learn.microsoft.com/dotnet/csharp/language-reference/compiler-messages/warning-waves) is already exhaustive for warning waves.

**Reply by huoyaoyuan** [reply]
The reason you are not seeing the warning is because it's suppressed by NUnit.Analyzers: https://github.com/nunit/nunit.analyzers/blob/master/src/nunit.analyzers/DiagnosticSuppressors/NonNullableFieldOrPropertyIsUninitializedSuppressor.cs

So it is in fact a Rider issue that not handling the suppressor.

**Reply by SRNissen** [reply]
> So [this page](https://learn.microsoft.com/dotnet/csharp/language-reference/compiler-messages/warning-waves) is already exhaustive for warning waves.

It is not.

For example, try finding the exact warning that instigated this thread, CS8618.

**Reply by huoyaoyuan** [reply]
Most warnings aren't controlled by any options or warning waves. CS8618 isn't in the page because it isn't in any warning level, or effectively level 0. They are just enabled unconditionally with the proper language feature. The linked ErrorFacts source code is source-of-truth.
